### ISSUE TITLE
zpool remove on mirrored logs fails silently

### ISSUE THREAD

**Issue by arthurzenika** [openzfs/zfs#1422]
On the setup described in :  https://github.com/zfsonlinux/zfs/issues/1420 we have a mirrored log ZIL :: 

[...]
        logs
          mirror-3                      ONLINE       0     0     0
            wwn-0x5001517bb2a7cca9      ONLINE       0     0     0
            wwn-0x5001517bb2a7acae      ONLINE       0     0     0
[...]

doing "zpool remove data mirror-3" doesn't complain, returns a 0 error code, but doesn't seem to work. 

Doing zpool status still shows the same thing. 

We tried to put the disks offline (zpool offline data wwn-0x5001517bb2a7acae) before the remove but this does not help. 

Is it possible that zpool status is showing wrong information (we don't have any disk activity on the ZIL, dstat -d -D sdw,sdx shows zeros) ? Does anyone have an idea on how to debug this ?

**Comment by wrouesnel** [comment]
I'm current affected by this on the current daily for Ubuntu on kernel 13.13. No matter what I try I can't actually remove the mirrored log devices. `zpool remove mirror-2` reports no errors but doesn't remove any devices.

```
  pool: storage
 state: DEGRADED
status: One or more devices has been taken offline by the administrator.
        Sufficient replicas exist for the pool to continue functioning in a
        degraded state.
action: Online the device using 'zpool online' or replace the device with
        'zpool replace'.
  scan: scrub repaired 6.50M in 32h44m with 0 errors on Fri Dec 19 16:34:14 2014
config:

        NAME                                              STATE     READ WRITE CKSUM
        storage                                           DEGRADED     0     0     0
          raidz3-0                                        ONLINE       0     0     0
            ata-ST2000DM001-1CH164_S1E29FS0               ONLINE       0     0     0
            ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2734507       ONLINE       0     0     0
            ata-WDC_WD20EZRX-00D8PB0_WD-WMC4M3194337      ONLINE       0     0     0
            ata-WDC_WD20EZRX-00DC0B0_WD-WMC301081511      ONLINE       0     0     0
            ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2734373       ONLINE       0     0     0
            ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2802365       ONLINE       0     0     0
            ata-ST2000DM001-1CH164_Z340PXSZ               ONLINE       0     0     0
            ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2882175       ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE358947      ONLINE       0     0     0
            ata-WDC_WD15EARS-00S8B1_WD-WCAVY4004247       ONLINE       0     0     0
          raidz3-1                                        ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE403743      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE401801      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE380254      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZA6592392      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE428280      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE411377      ONLINE       0     0     0
            ata-WDC_WD20EARX-00PASB0_WD-WCAZAE410176      ONLINE       0     0     0
            ata-ST2000DM001-9YN164_Z240453S               ONLINE       0     0     0
            ata-ST2000DM001-9YN164_Z1E0V4B7               ONLINE       0     0     0
            ata-WDC_WD20EZRX-00D8PB0_WD-WMC4M0DNECCF      ONLINE       0     0     0
        logs
          mirror-2                                        UNAVAIL      0     0     0  insufficient replicas
            ata-Corsair_Force_3_SSD_11436522000012071599  OFFLINE      0     0     0
            ata-Corsair_Force_3_SSD_11436522000012072283  OFFLINE      0     0     0
        cache
          ata-OCZ-AGILITY3_OCZ-PNYK3738U618C360           ONLINE       0     0     0

errors: No known data errors

```
storage:
    version: 28
    name: 'storage'
    state: 0
    txg: 33647724
    pool_guid: 18404054517318695605
    errata: 0
    hostid: 8323328
    hostname: 'xen'
    vdev_children: 3
    vdev_tree:
        type: 'root'
        id: 0
        guid: 18404054517318695605
        children[0]:
            type: 'raidz'
            id: 0
            guid: 9305773940725570617
            nparity: 3
            metaslab_array: 31
            metaslab_shift: 37
            ashift: 12
            asize: 15002873364480
            is_log: 0
            create_txg: 4
            children[0]:
                type: 'disk'
                id: 0
                guid: 14721724001840578894
                path: '/dev/disk/by-id/ata-ST2000DM001-1CH164_S1E29FS0-part1'
                whole_disk: 1
                DTL: 748120
                create_txg: 4
            children[1]:
                type: 'disk'
                id: 1
                guid: 14634396826649144688
                path: '/dev/disk/by-id/ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2734507-part1'
                whole_disk: 1
                DTL: 160228
                create_txg: 4
            children[2]:
                type: 'disk'
                id: 2
                guid: 7702592444643697601
                path: '/dev/disk/by-id/ata-WDC_WD20EZRX-00D8PB0_WD-WMC4M3194337-part1'
                whole_disk: 1
                DTL: 1255687
                create_txg: 4
            children[3]:
                type: 'disk'
                id: 3
                guid: 12472132331821042284
                path: '/dev/disk/by-id/ata-WDC_WD20EZRX-00DC0B0_WD-WMC301081511-part1'
                whole_disk: 1
                DTL: 484987
                create_txg: 4
            children[4]:
                type: 'disk'
                id: 4
                guid: 12111841628991602074
                path: '/dev/disk/by-id/ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2734373-part1'
                whole_disk: 1
                DTL: 160225
                create_txg: 4
            children[5]:
                type: 'disk'
                id: 5
                guid: 15322441750202816443
                path: '/dev/disk/by-id/ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2802365-part1'
                whole_disk: 1
                DTL: 160224
                create_txg: 4
            children[6]:
                type: 'disk'
                id: 6
                guid: 11040241086533342201
                path: '/dev/disk/by-id/ata-ST2000DM001-1CH164_Z340PXSZ-part1'
                whole_disk: 1
                DTL: 671473
                create_txg: 4
            children[7]:
                type: 'disk'
                id: 7
                guid: 12043741682904314628
                path: '/dev/disk/by-id/ata-WDC_WD15EARS-00Z5B1_WD-WMAVU2882175-part1'
                whole_disk: 1
                DTL: 160222
                create_txg: 4
            children[8]:
                type: 'disk'
                id: 8
                guid: 12982577528650286064
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE358947-part1'
                whole_disk: 1
                DTL: 66699
                create_txg: 4
            children[9]:
                type: 'disk'
                id: 9
                guid: 12183385382034234700
                path: '/dev/disk/by-id/ata-WDC_WD15EARS-00S8B1_WD-WCAVY4004247-part1'
                whole_disk: 1
                DTL: 160220
                create_txg: 4
        children[1]:
            type: 'raidz'
            id: 1
            guid: 14092802330132323438
            nparity: 3
            metaslab_array: 24286
            metaslab_shift: 37
            ashift: 9
            asize: 20003846881280
            is_log: 0
            create_txg: 16090928
            children[0]:
                type: 'disk'
                id: 0
                guid: 12615390902135672413
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE403743-part1'
                whole_disk: 1
                DTL: 24863
                create_txg: 16090928
            children[1]:
                type: 'disk'
                id: 1
                guid: 2667231458259228099
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE401801-part1'
                whole_disk: 1
                DTL: 24862
                create_txg: 16090928
            children[2]:
                type: 'disk'
                id: 2
                guid: 14398973086626766034
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE380254-part1'
                whole_disk: 1
                DTL: 24861
                create_txg: 16090928
            children[3]:
                type: 'disk'
                id: 3
                guid: 15477698036315389930
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZA6592392-part1'
                whole_disk: 1
                DTL: 24860
                create_txg: 16090928
            children[4]:
                type: 'disk'
                id: 4
                guid: 15800598362685821861
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE428280-part1'
                whole_disk: 1
                DTL: 24859
                create_txg: 16090928
            children[5]:
                type: 'disk'
                id: 5
                guid: 18164124165422162773
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE411377-part1'
                whole_disk: 1
                DTL: 24858
                create_txg: 16090928
            children[6]:
                type: 'disk'
                id: 6
                guid: 10430475747903180549
                path: '/dev/disk/by-id/ata-WDC_WD20EARX-00PASB0_WD-WCAZAE410176-part1'
                whole_disk: 1
                DTL: 24857
                create_txg: 16090928
            children[7]:
                type: 'disk'
                id: 7
                guid: 8720790458023090378
                path: '/dev/disk/by-id/ata-ST2000DM001-9YN164_Z240453S-part1'
                whole_disk: 1
                DTL: 24290
                create_txg: 16090928
            children[8]:
                type: 'disk'
                id: 8
                guid: 7702148493852147286
                path: '/dev/disk/by-id/ata-ST2000DM001-9YN164_Z1E0V4B7-part1'
                whole_disk: 1
                DTL: 24856
                create_txg: 16090928
            children[9]:
                type: 'disk'
                id: 9
                guid: 9572823555639542046
                path: '/dev/disk/by-id/ata-WDC_WD20EZRX-00D8PB0_WD-WMC4M0DNECCF-part1'
                whole_disk: 1
                DTL: 1626320
                create_txg: 16090928
        children[2]:
            type: 'mirror'
            id: 2
            guid: 6845648256434883313
            metaslab_array: 1686687
            metaslab_shift: 29
            ashift: 13
            asize: 60007383040
            is_log: 1
            removing: 1
            create_txg: 25111027
            children[0]:
                type: 'disk'
                id: 0
                guid: 2368769868690690475
                path: '/dev/disk/by-id/ata-Corsair_Force_3_SSD_11436522000012071599-part1'
                whole_disk: 1
                DTL: 1064888
                create_txg: 25111027
                offline: 1
            children[1]:
                type: 'disk'
                id: 1
                guid: 4971990348561151886
                path: '/dev/disk/by-id/ata-Corsair_Force_3_SSD_11436522000012072283-part1'
                whole_disk: 1
                DTL: 1064887
                create_txg: 25111027
                offline: 1
    features_for_read:

```

**Comment by behlendorf** [comment]
This should report an error because you're not including the pool name in the remove command.  Try `zpool remove storage mirror-2` to remove the devices.

**Comment by wrouesnel** [comment]
That's a typo in my complaint. The full command still fails silently.

On 17 January 2015 5:37:42 am AEDT, Brian Behlendorf <notifications@github.com> wrote:
>This should report an error because you're not including the pool name
>in the remove command.  Try `zpool remove storage mirror-2` to remove
>the devices.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/zfsonlinux/zfs/issues/1422#issuecomment-70300029

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by behlendorf** [comment]
OK good, could you grab the error code from the failed command.  It might provide some insight.

```
zpool remove storage mirror-2; echo $?
```

**Comment by wrouesnel** [comment]
I get code 0 back from `zpool remove storage mirror-2` but the mirror is not removed.

On 17 January 2015 8:48:26 am AEDT, Brian Behlendorf <notifications@github.com> wrote:
>OK good, could you grab the error code from the failed command.  It
>might provide some insight.
>
>```
>zpool remove storage mirror-2; echo $?
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/zfsonlinux/zfs/issues/1422#issuecomment-70327858

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by wrouesnel** [comment]
So I've tried b333z patch for my ZFS kernel, and I still can't get the log devices to actually remove - the same issue arises, the silent fail. As best I can guess, the issue is the mirror volume is unavailable, but the devices are just marked as offline so the checks to remove them still fail out.

One thing I noticed is that when I do an import, the log device isn't immediately known - maybe what's needed is a check to see if the log device could be removed on import?

**Comment by drakedevel** [comment]
As a heads up, I encountered the leaked SLOG space issue on a recently-created pool (made late last year on ZoL with whatever the latest ZoL was at the time, I'm afraid I don't have the version handy). I never had any sort of SLOG mirror, it was just a single device, added once at pool creation and not touched until I tried to remove it. I had to resort to @b333z 's now-ancient systemtap script (suitably adjusted as the original probe point no longer worked) to get the device to remove, but it appears that the hypothesized ancient bug causing the leak is back or was never actually fixed.
